cantera fails to build with Python 3.11: ValueError: invalid mode: 'rU'

Reported in full at https://bugzilla.redhat.com/show_bug.cgi?id=2094258

**Problem description**

cantera fails to build with Python 3.11.0b3.

**Steps to reproduce**


ValueError: invalid mode: 'rU':
  File "/builddir/build/BUILD/cantera-2.6.0/SConstruct", line 2020:
    SConscript("build/python/SConscript")
  File "/usr/lib/python3.11/site-packages/SCons/Script/SConscript.py", line 660:
    return method(*args, **kw)
  File "/usr/lib/python3.11/site-packages/SCons/Script/SConscript.py", line 597:
    return _SConscript(self.fs, *files, **subst_kw)
  File "/usr/lib/python3.11/site-packages/SCons/Script/SConscript.py", line 285:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "/builddir/build/BUILD/cantera-2.6.0/build/python/SConscript", line 98:
    setup_cfg = localenv.SubstFile("setup.cfg", "setup.cfg.in")
  File "/usr/lib/python3.11/site-packages/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/builddir/build/BUILD/cantera-2.6.0/site_scons/site_tools/subst.py", line 154:
    return env.SubstGeneric(target,
  File "/usr/lib/python3.11/site-packages/SCons/Environment.py", line 238:
    return super().__call__(target, source, *args, **kw)
  File "/usr/lib/python3.11/site-packages/SCons/Util.py", line 742:
    return self.method(*nargs, **kwargs)
  File "/usr/lib/python3.11/site-packages/SCons/Builder.py", line 662:
    return self._execute(env, target, source, OverrideWarner(kw), ekw)
  File "/usr/lib/python3.11/site-packages/SCons/Builder.py", line 568:
    tlist, slist = self._create_nodes(env, target, source)
  File "/usr/lib/python3.11/site-packages/SCons/Builder.py", line 530:
    target, source = self.emitter(target=tlist, source=slist, env=env)
  File "/builddir/build/BUILD/cantera-2.6.0/site_scons/site_tools/subst.py", line 124:
    keys = _subst_keys(path, pattern)
  File "/builddir/build/BUILD/cantera-2.6.0/site_scons/site_tools/subst.py", line 49:
    f = open(source, "rU")
error: Bad exit status from /var/tmp/rpm-tmp.S5BgPF (%build)

open(), io.open(), codecs.open() and fileinput.FileInput no longer accept 'U' (“universal newline”) in the file mode. This flag was deprecated since Python 3.3. In Python 3, the “universal newline” is used by default when a file is open in text mode. The newline parameter of open() controls how universal newlines works. (Contributed by Victor Stinner in bpo-37330.)

https://bugs.python.org/issue?@action=redirect&bpo=37330
https://docs.python.org/3.11/whatsnew/3.11.html

For the build logs, see:
https://copr-be.cloud.fedoraproject.org/results/@python/python3.11/fedora-rawhide-x86_64/04495077-cantera/

For all our attempts to build cantera with Python 3.11, see:
https://copr.fedorainfracloud.org/coprs/g/python/python3.11/package/cantera/

Testing and mass rebuild of packages is happening in copr. You can follow these instructions to test locally in mock if your package builds with Python 3.11:
https://copr.fedorainfracloud.org/coprs/g/python/python3.11/

Let us know here if you have any questions.

Python 3.11 is planned to be included in Fedora 37. To make that update smoother, we're building Fedora packages with all pre-releases of Python 3.11.
A build failure prevents us from testing all dependent packages (transitive [Build]Requires), so if this package is required a lot, it's important for us to get it fixed soon.
We'd appreciate help from the people who know this package best, but if you don't want to work on this now, let us know so we can try to work around it on our side.

**System information**

- Cantera version: 2.6.0 release
- OS: Fedora Linux 37/rawhide
- Python/MATLAB/other software versions: Python 3.11b

I fixed the above error in (draft) PR #1326, but then we get another error:

```
/usr/bin/python3 -c "import Cython.Build; Cython.Build.cythonize('build/python/cantera/_cantera.pyx')"
Compiling build/python/cantera/_cantera.pyx because it changed.
[1/1] Cythonizing build/python/cantera/_cantera.pyx
g++ -o build/temp-py/_cantera.os -c -std=c++11 -pthread -O3 -Wno-inline -g -fPIC -DNDEBUG -Iinclude -I/usr/include/python3.11 -I/usr/lib64/python3.11/site-packages/numpy/core/include -I/usr/include/eigen3 build/python/cantera/_cantera.cpp
In file included from /usr/lib64/python3.11/site-packages/numpy/core/include/numpy/ndarraytypes.h:1960,
                 from /usr/lib64/python3.11/site-packages/numpy/core/include/numpy/ndarrayobject.h:12,
                 from /usr/lib64/python3.11/site-packages/numpy/core/include/numpy/arrayobject.h:5,
                 from build/python/cantera/_cantera.cpp:749:
/usr/lib64/python3.11/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: #warning "Using deprecated NumPy API, disable it with " "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-Wcpp]
   17 | #warning "Using deprecated NumPy API, disable it with " \
      |  ^~~~~~~
In file included from /usr/include/python3.11/Python.h:38,
                 from build/python/cantera/_cantera.cpp:20:
/usr/include/python3.11/pyport.h: In instantiation of 'type {anonymous}::_Py_CAST_impl(const expr_type&) [with type = _object*; expr_type = int]':
build/python/cantera/_cantera.cpp:193144:9:   required from here
/usr/include/python3.11/pyport.h:47:24: error: invalid 'static_cast' from type 'int' to type '_object*'
   47 |                 return static_cast<type>(const_cast<expr_type &>(expr));
      |                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
scons: *** [build/temp-py/_cantera.os] Error 1
scons: building terminated because of errors.
```

--- a/.git-blame-ignore-revs
+++ b/.git-blame-ignore-revs
@@ -1,0 +1,3 @@
+# .git-blame-ignore-revs
+# reformat (black) subst.py during update to Python 3.11
+c3b5d2847a1cfb606257c1881932dc50eb30cabf

--- a/site_scons/site_tools/subst.py
+++ b/site_scons/site_tools/subst.py
@@ -28,7 +28,7 @@ def _subst_file(target, source, env, pattern, replace):
         if not SCons.Util.is_String(value):
             raise SCons.Errors.UserError("Substitution must be a string.")
         return value
-    #print 'pattern  = ' , pattern
+
     contents = re.sub(pattern, subfn, contents)
 
     # Write file
@@ -38,6 +38,7 @@ def _subst_file(target, source, env, pattern, replace):
     finally:
         f.close()
 
+
 # Determine which keys are used
 def _subst_keys(source, pattern):
     # Read file
@@ -45,16 +46,17 @@ def _subst_keys(source, pattern):
 
     # Determine keys
     keys = []
+
     def subfn(mo):
         key = mo.group("key")
         if key:
             keys.append(key)
-        return ''
+        return ""
 
-    
     re.sub(pattern, subfn, contents)
 
     return keys
+
 
 # Get the value of a key as a string, or None if it is not in the environment
 def _subst_value(env, key):
@@ -66,17 +68,10 @@ def _subst_value(env, key):
 
     # env.subst already returns a string even if it is stored as a number
     # such as env['HAVE_XYZ'] = 1
-    #print 'key = ', key
-    #print '  straight env = ', env[key]
-    #print '  str of the thing = ', str(env[key])
-    #print '  subst(${}) of the thing = ', env.subst("${%s}" % key) 
-    #print '  %s of the thing = ', "%s" % str(env[key])
     aa = env[key]
     if aa == []:
-       aa = ''
+        aa = ""
     return aa
-    #return str(env[key])
-    #return env.subst("${%s}" % key)
 
 
 # Builder related functions
@@ -87,19 +82,21 @@ def _subst_action(target, source, env):
     # Substitute in the files
     pattern = env["SUBST_PATTERN"]
     replace = env["SUBST_REPLACE"]
-    #print 'SUBSTITUTE: ', pattern, ' for ', replace
 
     for (t, s) in zip(target, source):
         _subst_file(str(t), str(s), env, pattern, replace)
 
     return 0
 
+
 # Builder message
 def _subst_message(target, source, env):
-    items = ["Substituting vars from %s to %s" % (s, t)
-             for (t, s) in zip(target, source)]
+    items = [
+        "Substituting vars from %s to %s" % (s, t) for (t, s) in zip(target, source)
+    ]
 
     return "\n".join(items)
+
 
 # Builder dependency emitter
 def _subst_emitter(target, source, env):
@@ -117,7 +114,6 @@ def _subst_emitter(target, source, env):
         d = dict()
         for key in keys:
             value = _subst_value(env, key)
-            # print 'key = ', key, ' -> value = ', value
             if not value is None:
                 d[key] = value
 
@@ -131,6 +127,8 @@ def _subst_emitter(target, source, env):
 ##############################################################################
 
 _SubstFile_pattern = "@(?P<key>\w*?)@"
+
+
 def _SubstFile_replace(env, mo):
     key = mo.group("key")
     if not key:
@@ -141,11 +139,14 @@ def _SubstFile_replace(env, mo):
         raise SCons.Errors.UserError("Error: key %s does not exist" % key)
     return value
 
+
 def SubstFile(env, target, source):
-    return env.SubstGeneric(target,
-                            source,
-                            SUBST_PATTERN=_SubstFile_pattern,
-                            SUBST_REPLACE=_SubstFile_replace)
+    return env.SubstGeneric(
+        target,
+        source,
+        SUBST_PATTERN=_SubstFile_pattern,
+        SUBST_REPLACE=_SubstFile_replace,
+    )
 
 
 # A substitutor similar to config.h header substitution
@@ -167,7 +168,11 @@ def SubstFile(env, target, source):
 # other defines that you do not desire to be replaced.
 ##############################################################################
 
-_SubstHeader_pattern = "(?m)^(?P<space>\\s*?)(?P<type>#define|#undef)\\s+?@(?P<key>\w+?)@(?P<ending>.*?)$"
+_SubstHeader_pattern = (
+    "(?m)^(?P<space>\\s*?)(?P<type>#define|#undef)\\s+?@(?P<key>\w+?)@(?P<ending>.*?)$"
+)
+
+
 def _SubstHeader_replace(env, mo):
     space = mo.group("space")
     type = mo.group("type")
@@ -192,18 +197,20 @@ def _SubstHeader_replace(env, mo):
     # It was #undef
     return "%s#undef %s" % (space, key)
 
+
 def SubstHeader(env, target, source):
-    return env.SubstGeneric(target,
-                            source,
-                            SUBST_PATTERN=_SubstHeader_pattern,
-                            SUBST_REPLACE=_SubstHeader_replace)
+    return env.SubstGeneric(
+        target,
+        source,
+        SUBST_PATTERN=_SubstHeader_pattern,
+        SUBST_REPLACE=_SubstHeader_replace,
+    )
 
 
 def generate(env, **kw):
     # The generic builder
     subst = SCons.Action.Action(_subst_action, _subst_message)
-    env['BUILDERS']['SubstGeneric'] = Builder(action=subst,
-                                              emitter=_subst_emitter)
+    env["BUILDERS"]["SubstGeneric"] = Builder(action=subst, emitter=_subst_emitter)
 
     # Additional ones
     env.AddMethod(SubstFile)

--- a/site_scons/site_tools/subst.py
+++ b/site_scons/site_tools/subst.py
@@ -12,7 +12,7 @@ import re
 
 from SCons.Script import *
 import SCons.Errors
-
+from pathlib import Path
 
 # Helper/core functions
 ##############################################################################
@@ -20,12 +20,7 @@ import SCons.Errors
 # Do the substitution
 def _subst_file(target, source, env, pattern, replace):
     # Read file
-    #print 'CALLING SUBST_FILE'
-    f = open(source, "rU")
-    try:
-        contents = f.read()
-    finally:
-        f.close()
+    contents = Path(source).read_text()
 
     # Substitute, make sure result is a string
     def subfn(mo):
@@ -46,11 +41,7 @@ def _subst_file(target, source, env, pattern, replace):
 # Determine which keys are used
 def _subst_keys(source, pattern):
     # Read file
-    f = open(source, "rU")
-    try:
-        contents = f.read()
-    finally:
-        f.close()
+    contents = Path(source).read_text()
 
     # Determine keys
     keys = []

